attribute-masking-service/src/main/java/uk/gov/gchq/palisade/service/attributemask/stream/config/AkkaRunnableGraph.java

```diff
@@ -69,48 +74,57 @@ KafkaProducerService kafkaProducerService(
         return source
                 // Extract token from message, keeping track of original message
                 .map(committableMessage -> new Pair<>(committableMessage, new String(committableMessage.record().headers().lastHeader(Token.HEADER).value(), Charset.defaultCharset())))
 
                 // Store authorised request in persistence, keeping track of original message and token
-                .mapAsync(PARALLELISM, (Pair<CommittableMessage<String, AttributeMaskingRequest>, String> messageAndToken) -> {
-                    AttributeMaskingRequest request = messageAndToken.first().record().value();
-                    return service.storeAuthorisedRequest(messageAndToken.second(), request)
-                            .thenApply(ignored -> new Tuple3<>(messageAndToken.first(), messageAndToken.second(), request));
-                })
-
-                // Mask resource attributes, keeping track of original message and token
-                .map(messageTokenRequest -> new Tuple3<>(messageTokenRequest.t1(), messageTokenRequest.t2(), service.maskResourceAttributes(messageTokenRequest.t3())))
-
-                // Build producer record, copying the partition, keeping track of original message
-                .map((Tuple3<CommittableMessage<String, AttributeMaskingRequest>, String, AttributeMaskingResponse> messageTokenResponse) -> {
-                    ConsumerRecord<String, AttributeMaskingRequest> requestRecord = messageTokenResponse.t1().record();
-                    // In the future, consider recalculating the token according to number of upstream/downstream partitions available
-                    return new Pair<>(messageTokenResponse.t1(), new ProducerRecord<>(outputTopic.getName(), requestRecord.partition(), requestRecord.key(), messageTokenResponse.t3(), requestRecord.headers()));
+                .mapAsync(PARALLELISM, (Pair<CommittableMessage<String, AttributeMaskingRequest>, String> messageAndToken) ->
+                    service.storeAuthorisedRequest(messageAndToken.second(), messageAndToken.first().record().value())
+                            .thenApply(auditable -> Pair.create(messageAndToken.first(), auditable))
+                )
+
+                // Mask resource attributes, keeping track of original message
+                .map(request -> Pair.create(request.first(), service.maskResourceAttributes(request.second().getAttributeMaskingRequest())
+                        .chain(request.second().getAuditErrorMessage())))
+
+                // Build producer message, copying the partition, keeping track of original message
+                .map((Pair<CommittableMessage<String, AttributeMaskingRequest>, AuditableAttributeMaskingResponse> response) -> {
+                    ConsumerRecord<String, AttributeMaskingRequest> requestRecord = response.first().record();
+                    return Optional.ofNullable(response.second().getAuditErrorMessage()).map(audit ->
```

Comment:
This makes me think the AuditableAttributeMaskingResponse is a strong contender for an `Either` type
Something like https://stackoverflow.com/questions/26162407/is-there-an-equivalent-of-scalas-either-in-java-8/26164155#26164155 or even https://www.vavr.io/ (probably a bit overkill)?

Reply:
Considering there are only two of these in this service & in others they may only be one it is most definitely overkill. Worst still is if it is then used as either.left == null etc. we have lost static typing checks that will catch human errors.